### ISSUE TITLE
Drop dependency on Media Element

### ISSUE THREAD

**Issue by kornelski** [ResponsiveImagesCG/picture-element#62]
In response to objection that `<picture>`'s handling of `<source>` is painfully complicated:

http://lists.w3.org/Archives/Public/public-html/2013Sep/0137.html

I've reviewed the tests:

http://w3c-test.org/web-platform-tests/submissions/326/html/semantics/embedded-content-0/media-elements/loading-the-media-resource/

and found that significant number of them is testing `MediaElement`'s `networkState` property.

`MediaElement` is complicated due to all the states needed to expose loading and buffering states for video. Images don't need that, so this burden could be dropped.

Second contributor to number of tests (edge cases) seems to be `<source onerror>`.  A lot of those cases could be avoided if all error events were fired on `<picture>`.

**Comment by nwtn** [comment]
As it stands, `<picture>` already avoids `MediaElement`'s `networkState`. Firing errors at `<picture>` instead of `<source>` is an interesting idea, but I'm not sure of the implications.

**Comment by anselmh** [comment]
I'm not sure but doesn't it cause confusing issues when you fire all errors on picture? Let's say you're doing content negotiation (different image formats) it would be helpful to get the error on the one that caused the error not only some message on the wrapper?

**Comment by marcoscaceres** [comment]
I've removed dependency on Media Elements. I think it's fine to fire an error on the picture element - but I'm going to defer error handling for now.

**Comment by kornelski** [comment]
Hailing @zcorpan to give feedback on this.

**Comment by zcorpan** [comment]
The inherent complexity comes from using multiple elements. Changing the algorithm doesn't remove the possible cases that come from using multiple elements. My recommendation is what I said in the original email: use a single element.

**Comment by marcoscaceres** [comment]
@zcorpan I don't see how that can happen without ending up with uglier markup than we already got (declaring a bunch of stuff in attributes is going to be equally as hard to parse). Quite a few people have iterated over and over possible markup solutions (e.g., srcset is an example of how ugly it gets). Solving it with HTTP (Client Hints) was also problematic... we are kinda stuck.

**Comment by kornelski** [comment]
[On www-style I've written first draft of simplified source selection algorithm](http://lists.w3.org/Archives/Public/www-style/2013Sep/0357.html)
- algorithm is supposed to be stateless and idempotent, so UA can re-run it as often as it likes.
- selection should always be done asynchronously to appear atomic to JS (if JS adds/removes `<source>` elements the algorithm should be scheduled to run on the next tick, so JS can't mess with the sources while browser is evaluating them).
- `null` return from the algorithm below means no image displayed (same as `<img>` with no `src`).
- once `<source>` node is selected here, browser should evaluate `src` and `srcset` as usual.

This is JS pseudocode.

```
function sourceSelection(picture) {
    for(var i=0; i < picture.childNodes.length; i++) {
        var child = picture.childNodes[i];
        if (child.tagName != 'SOURCE' || child.namespaceURI != xhtml_namespace)  continue;

        if (child.getAttribute('type')) {
            if (!isSupportedMimeType(child.getAttribute('type'))) continue;
        }

        if (child.getAttribute('media')) {
            var mq = matchMedia(child.getAttribute('media'));
            mq.addEventListener(sourceSelection); // re-run selection algorithm  whenever MQ changes

            // <optional> to support async CSS parsing (external stylesheets)
            if (!CSSisLoaded && mq.cannotFullyEvaluateYet) {                                
                addEventListener('CSSisLoaded', sourceSelection); // re-run selection algorithm when CSS is loaded
                return null;
            }
            // </optional>

            if (!mq.matches) continue;
        }
        return child;
    }
    return null;
}
```

Basically chooses first child `<source>` element that matches on type and media (omitted `type`/`media` attributes match).

The twist is that when browser encounters media query it can't evaluate (indicated by hypothetical `mq.cannotFullyEvaluateYet` property, e.g. when `min-width:` is used in an `<iframe>` that has no layout yet or uses a variable #64 that is defined in external stylesheet) it aborts the selection and re-tries it when MQ changes or browser starts first render (indicated by hypothetical `CSSisLoaded` variable and event).

**Comment by tabatkins** [comment]
I think this is handled now in the spec? @pornel can you verify?

**Comment by kornelski** [comment]
Yes, it's good now!
